File: packages/r/openproblems.docs/R/read_component_spec.R
#' Read component spec
#'
#' This function reads a component spec from a yaml file.
#'
#' @param path Path to a component spec yaml, usually in `src/api/comp_*.yaml`
#' @return A list with compontent info and arguments
#'
#' @export
#' @examples
#' path <- system.file(
#'   "extdata", "example_project", "api", "comp_method.yaml",
#'   package = "openproblems.docs"
#' )
#'
#' read_component_spec(path)
read_component_spec <- function(path) {
  spec_yaml <- openproblems::read_nested_yaml(path)
  list(
    info = read_component_spec__process_info(spec_yaml, path),
    args = read_component_spec_arguments(spec_yaml, path)
  )
}

#' @importFrom openproblems.utils list_as_data_frame is_list_a_dataframe
read_component_spec__process_info <- function(spec_yaml, path) { # nolint object_length_linter
  df <- list_as_data_frame(spec_yaml)
  if (is_list_a_dataframe(spec_yaml$info)) {
    df <- dplyr::bind_cols(df, list_as_data_frame(spec_yaml$info))
  }
  if (is_list_a_dataframe(spec_yaml$info$type_info)) {
    df <- dplyr::bind_cols(df, list_as_data_frame(spec_yaml$info$type_info))
  }
  df$file_name <- basename(path) |> str_replace_all("\\.yaml", "")
  as_tibble(df)
}

read_component_spec_arguments <- function(spec_yaml, path) {
  arguments <- spec_yaml$arguments
  for (arg_group in spec_yaml$argument_groups) {
    arguments <- c(arguments, arg_group$arguments)
  }
  map_dfr(arguments, function(arg) {
    df <- list_as_data_frame(arg)
    if (is_list_a_dataframe(arg$info)) {
      df <- dplyr::bind_cols(df, list_as_data_frame(arg$info))
    }
    df$file_name <- basename(path) |> str_replace_all("\\.yaml", "")
    df$arg_name <- str_replace_all(arg$name, "^-*", "")
    df$direction <- df$direction %||% "input" %|% "input"
    df$parent <- df$`__merge__` %||% NA_character_ |>
      basename() |>
      str_replace_all("\\.yaml", "")
    df$required <- df$required %||% FALSE %|% FALSE
    df$default <- df$default %||% NA_character_ |> as.character()
    df$example <- df$example %||% NA_character_ |> as.character()
    df$description <- df$description %||% NA_character_ |> as.character()
    df$summary <- df$summary %||% NA_character_ |> as.character()
    as_tibble(df)
  })
}

File: packages/python/openproblems/src/openproblems/project/docs/read_component_spec.py
from __future__ import annotations
import os
import re


def read_component_spec(path: str) -> dict:
    """Read a component spec from a ``comp_*.yaml`` file.

    Args:
        path: Path to a component spec yaml (usually ``src/api/comp_*.yaml``).

    Returns:
        A dict with keys ``info`` (dict) and ``args`` (list of dicts).
    """
    from .. import read_nested_yaml

    data = read_nested_yaml(path)
    return {
        "info": _process_info(data, path),
        "args": _process_arguments(data, path),
    }


def _process_info(data: dict, path: str) -> dict:
    file_name = re.sub(r"\.ya?ml$", "", os.path.basename(path))
    info: dict = {"file_name": file_name}

    # Top-level fields
    for key in ("label", "summary", "description", "namespace"):
        info[key] = data.get(key)

    # Merge info block (may override Nones above)
    for key, val in (data.get("info") or {}).items():
        if info.get(key) is None:
            info[key] = val

    # Merge info.type_info
    for key, val in ((data.get("info") or {}).get("type_info") or {}).items():
        if info.get(key) is None:
            info[key] = val

    return info


def _process_arguments(data: dict, path: str) -> list[dict]:
    file_name = re.sub(r"\.ya?ml$", "", os.path.basename(path))

    arguments = list(data.get("arguments") or [])
    for arg_group in data.get("argument_groups") or []:
        arguments.extend(arg_group.get("arguments") or [])

    result = []
    for arg in arguments:
        arg_info = arg.get("info") or {}
        merge_ref = arg.get("__merge__")
        parent = (
            re.sub(r"\.ya?ml$", "", os.path.basename(merge_ref)) if merge_ref else None
        )

        default = arg.get("default")
        example = arg.get("example")
        if isinstance(example, list):
            example = example[0] if example else None

        result.append(
            {
                "file_name": file_name,
                "arg_name": re.sub(r"^-+", "", arg.get("name", "")),
                "type": arg.get("type", ""),
                "direction": arg.get("direction") or "input",
                "required": bool(arg.get("required"))
                if arg.get("required") is not None
                else False,
                "default": str(default) if default is not None else None,
                "example": str(example) if example is not None else None,
                "description": arg.get("description") or arg_info.get("description"),
                "summary": arg.get("summary") or arg_info.get("summary"),
                "parent": parent,
            }
        )

    return result
